### ISSUE TITLE
Is Bayes HPO possible?

### ISSUE THREAD

**Issue by Marmeladenbrot** [facebook/prophet#1674]
I would like to find the best hyper parameter set for my forecast model.

At the moment I'm doing a grid search which takes ages even on a really beefy machine with multiprocess.
I would like to use Bayes HPO to converge to a maximum over time but haven't found a way to do so (like hyperopt with LightGBM).

Is there a known way to achieve this?

**Comment by amw5g** [comment]
I've seen others use `hyperopt` with fbprophet, like #1320.  I've also successfully integrated it with `optuna`.  
e.g.,
```
def prophet_objective(trial, df_past):
    seasonality_mode = trial.suggest_categorical('seasonality_mode', ["additive", "multiplicative"]) #
    changepoint_prior_scale = trial.suggest_uniform('changepoint_prior_scale', 0.01,0.2)
    changepoint_range = trial.suggest_float('changepoint_range', 0.50,0.85)
    seasonality_prior_scale = trial.suggest_float('seasonality_prior_scale', 0.01,10, log=True)#
    growth = trial.suggest_categorical('growth', ["linear", "logistic"])
    holidays_subset = holidays.copy()
    m = Prophet(
                growth = growth,
                seasonality_mode = seasonality_mode,
                changepoint_prior_scale = changepoint_prior_scale,
                changepoint_range = changepoint_range,
                seasonality_prior_scale = seasonality_prior_scale,
                holidays= holidays_subset,
                uncertainty_samples=0
                )
    m.add_regressor('regressor1', mode='additive')
    m.add_regressor('regressor2', mode='additive')
    m.fit(df_past)
    #Using INITIAL days to train, build a forecast of HORIZON days long, every PERIOD days
    df_cv = cross_validation(m, initial='28 days', period='3 days', horizon = '5 days')
    df_p = performance_metrics(df_cv, rolling_window=1)
    mape = np.mean(df_p.mape)

    return mape

# some code to create df_past goes here
study = optuna.create_study(direction="minimize")
study.optimize(lambda trial: prophet_objective(trial, df_past), n_trials=32)    
print(study.best_params)
```

**Comment by bletham** [comment]
I just wanted to add a couple comments. The first is that if cross validation is taking a long time, look at how many cutoffs you are using and maybe you don't need that many. Since the forecast errors are correlated across cutoffs with overlapping horizons, the standard error of the performance metric (MAPE, MSE, whatever) will decrease at a rate worse than 1/sqrt(n) for n cutoffs. So the marginal gains of cutoffs with overlapping horizons can be quite low and maybe not worth the compute time.

The second is just to point out the recommendations in https://facebook.github.io/prophet/docs/diagnostics.html#hyperparameter-tuning for anyone who hasn't seen them yet. There are some parameters that most likely don't need to be tuned, which reduces the HPO search space and can further speed things up.
